### ISSUE TITLE
API: Enhance permissions to allow filtering by environments

### ISSUE THREAD

**Issue by rajdip-b** [keyshade-xyz/keyshade#260]
## Description

Right now, the roles that are created can have one or many projects associated with it. Associating projects mean that members who will have this role attached to them, will be able to perform the `authorities` in `WorkspaceRole` entity [refer to prisma schema].

We would like to have another layer of security in here. We would also like to introduce environment specific access, so that members can have access to only a specific set of environments in the project set by the admin.

Use case:

Consider that there's a project that has 3 environments - dev, stage and prod. It will be ideal to allow the developers access the `dev` and `stage` environments (and the secrets and variables in it) while the `prod` environment is only accessible by the DevOps team or such. This is where this feature will be helpful

## Solution

- Update the [`ProjectWorkspaceRoleAssociation`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/prisma/schema.prisma#L324) entity to include `environments` - specifying the environments accessible by the member.
- Create a reverse relation from [`Environment`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/prisma/schema.prisma#L271) on to [`ProjectWorkspaceRoleAssociation`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/prisma/schema.prisma#L324).
- Add a util function named `getCollectiveEnvironmentAuthorities`. You can take the reference from any [`get-collective-project-authorities.ts`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/common/get-collective-project-authorities.ts). The purpose of this function would be this: given the userId, project, and environment, it would need to fetch all the authorities that the member has over this environment.
- Update [`AuthorityCheckerService#checkAuthorityOverEnvironment`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/common/authority-checker.service.ts#L198) to incorporate the `getCollectiveEnvironmentAuthorities` function and filter the roles. 
- Update the [`projectIds`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/workspace-role/dto/create-workspace-role/create-workspace-role.ts#L22) to be a of type `Map<String, Array<String>>`. The association will be something like this: For every projectId, list of environmentIds that will be available to the role.
- Update the [`WorkspaceRoleService`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/workspace-role/service/workspace-role.service.ts) functions acordingly
- Update the required tests in `workspace-role` module.

**Comment by rajdip-b** [comment]
Hey @rayaanoidPrime! Could you hold up to this for a day or two? I'm making a major refactor to the way secrets, variables and environments are organized. I would notify you once its done.

**Comment by rayaanoidPrime** [comment]
Oh sure no worries

**Comment by rajdip-b** [comment]
Hey bro! @rayaanoidPrime 

v2.0.0 just got released! You can start working now. Do hit me up if you want help.

**Comment by rajdip-b** [comment]
Howdy @rayaanoidPrime, any updates yet?

**Comment by rayaanoidPrime** [comment]
Hey @rajdip-b sorry for being unresponsive just been kinda busy last few months. I remember being done with it and I think only the tests are left to be written. Let me see if I can do a pr this afternoon.

**Comment by rajdip-b** [comment]
No rush bro! Just let us know when you can put up the PR.

**Comment by rayaanoidPrime** [comment]
Hi! In light of the evolutions and additions that the code has undergone in the time after this issue was posted I have thought of a few changes to the initial solution format. (the core logic remains the same ). Let me know your thoughts
 - 
1. Update CreateWorkspaceRole DTO:
```bash
class CreateWorkspaceRole {
  // ... other properties ...
  @IsOptional()
  @ValidateNested({ each: true })
  @Type(() => ProjectEnvironments)
  readonly projectEnvironments?: ProjectEnvironments[]
}
class ProjectEnvironments {
  @IsString()
  projectSlug: string
  @IsArray()
  @IsString({ each: true })
  environmentSlugs: string[]
}
```
3. Modify WorkspaceRoleService :: createWorkspaceRole and updateWorkspaceRole methods to handle new project-environment associations.
4. Revise workspace-role.e2e.spec.ts:
-  Update existing tests and add new ones to cover the new functionality.
- Create actual environments (Dev and Prod) for each project in the test setup.
- Use these created environments in the tests instead of undefined environment names.
5. Update `checkAuthorityOverEnvironment` to incorporate the `getCollectiveEnvironmentAuthorities` function and filter the roles.
6. Add new helper function in WorkspaceService:
   - Create a new private method `getEnvironmentSlugToIdMap` to fetch environment IDs:
   ```bash
   private async getEnvironmentSlugToIdMap(environmentSlugs: string[]) {
     const environments = await this.prisma.environment.findMany({
       where: {
         slug: {
           in: environmentSlugs
         }
       }
     })
     return new Map(environments.map((env) => [env.slug, env.id]))
   }
   ```
   This function takes an array of environment slugs, queries the database to find matching environments, and returns a Map with environment slugs as keys and their corresponding IDs as values.

**Comment by rajdip-b** [comment]
Yes! I believe that's what needs to be done. Thanks for the detailed writeup.

**Comment by rajdip-b** [comment]
@rayaanoidPrime hey bro long time no see. I'm unassigning this for now. Feel free to pick this up later on.
